File: c14_graph_dfs_bfs/L130-m.cpp
//
// Created by LiHeng on 2020/7/5.
//


//从一个节点出发，使用 DFS 对一个图进行遍历时，能够遍历到的节点都是从初始节点可达的，DFS 常用来求解这种 可达性 问题。
//在程序实现 DFS 时需要考虑以下问题：
//    栈：用栈来保存当前节点信息，当遍历新节点返回时能够继续遍历当前节点。可以使用递归栈。
//    标记：和 BFS 一样同样需要对已经遍历过的节点进行标记。


// 这道题转化成了找到所有和边界为’O‘的元素的连通区域，这些区域是不能被更改成’X'的.

// 下面的解法仿照L79 visited数组作用类似之前写回溯算法中的当前记录数组的作用，
// 即发挥栈的作用，只是用了bool来作用，没有用压入弹出操作注意区别

#include <iostream>
#include <vector>
using namespace std;
class Solution{
private:
    int d[4][2] = {{-1,0},{0, 1},{1,0},{0,-1}} ;
    int m,n; //因为inArea里面用到了
    vector<vector<bool>> visited ;

    bool inArea(int x , int y){
        return x>=0 && x<m && y>=0 && y<n ;
    }

    void dfs(vector<vector<char>> &board,  int startx, int starty){
        if(board[startx][starty] != 'O')
            return;
        //if(!visited[startx][starty]) 就可以了，不需要判断board[startx][starty] == 'O'
        if(board[startx][starty] == 'O' && !visited[startx][starty]) {
            visited[startx][starty] = true ;
            board[startx][starty] = '#';
            //从[startx][starty]出发，4个方向寻找
            for(int i = 0 ; i < 4 ; i++){
                int newx = startx+d[i][0];
                int newy = starty+d[i][1];
                if(inArea(newx,newy))
                    dfs(board, newx, newy);
            }
            visited[startx][starty] = false ;
        }
    }

public:
    void solve(vector<vector<char>>& board) {
        m = board.size() ;
        if (m == 0) return;
        n = board[0].size();
        visited = vector<vector<bool>> (m, vector<bool>(n, false));
        for(int i=0 ; i<m ; ++i)
            for(int j=0 ; j<board[i].size() ; ++j) {
                bool isEdge = (i == 0 || j == 0 || i == m - 1 || j == n - 1);
                if(isEdge && board[i][j] == 'O')
                    dfs(board,i,j);
            }
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
                if (board[i][j] == '#') {
                    board[i][j] = 'O';
                }
            }
        }

    }

};

File: python-algorithm/algo_08_graph/graph_2_dfs/L130-m.py
#--coding:utf-8--


class Solution:
    def __init__(self, **kwargs):
        self.m = 0
        self.n = 0
        self.ori = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    def inArea(self, x, y):
        return x >= 0 and x < self.m and y >= 0 and y < self.n

    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        self.m = len(board)
        if self.m == 0:
            return board
        self.n = len(board[0])
        for i in range(self.m):
            for j in range(self.n):
                if (i == 0 or j == 0 or i == self.m - 1 or j == self.n - 1) and board[i][j] == 'O':
                    self.dfs(board, i, j)
        for i in range(self.m):
            for j in range(self.n):
                if board[i][j] == 'O':
                    board[i][j] = 'X'
                if board[i][j] == '#':
                    board[i][j] = 'O'

    def dfs(self, board, x, y):
        if board[x][y] != 'O':
            return
        if board[x][y] == 'O':
            board[x][y] = '#' # 已经可以防止4个方向中的回头路方向了
            for i in range(4):
                new_x = x + self.ori[i][0]
                new_y = y + self.ori[i][1]
                if self.inArea(new_x, new_y):
                    self.dfs(board, new_x, new_y)




# 额外写访问数组，其实没必要
class Solution1:
    def __init__(self, **kwargs):
        self.m = 0
        self.n = 0
        self.ori = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        self.flag = []

    def inArea(self, x, y):
        return x >= 0 and x < self.m and y >= 0 and y < self.n

    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        self.m = len(board)
        if self.m == 0:
            return board
        self.n = len(board[0])
        self.flag = [[False for j in range(self.n)] for i in range(self.m)]
        for i in range(self.m):
            for j in range(self.n):
                if (i == 0 or j == 0 or i == self.m - 1 or j == self.n - 1) and board[i][j] == 'O':
                    self.dfs(board, i, j)
        for i in range(self.m):
            for j in range(self.n):
                if board[i][j] == 'O':
                    board[i][j] = 'X'
                if board[i][j] == '#':
                    board[i][j] = 'O'

    def dfs(self, board, x, y):
        if board[x][y] != 'O':
            return
        if board[x][y] == 'O':
            self.flag[x][y] = True
            board[x][y] = '#'
            for i in range(4):
                new_x = x + self.ori[i][0]
                new_y = y + self.ori[i][1]
                if self.inArea(new_x, new_y) and not self.flag[new_x][new_y]:
                    self.dfs(board, new_x, new_y)
            self.flag[x][y] = False



